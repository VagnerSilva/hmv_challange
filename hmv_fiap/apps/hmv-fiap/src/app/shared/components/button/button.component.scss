button {
	width: 100%;
	min-width: 220px;
	height: 40px;
  text-transform: uppercase;

	@screen tbt {
		min-width: 200px;
		width: 100%;
		height: 42px;
	}
}

.hmv-icon--left,
.hmv-icon--rigth {
	position: relative;
	top: -2px;
}

.hmv-btn-primary {
	@apply bg-primary-pure
  text-gray-50;
}

.hmv-btn-primary--hover {
	@apply hover:bg-gray-50
  hover:text-primary-pure;
}

.hmv-btn-accent {
	@apply bg-gray-50
  text-primary-pure;
}

.hmv-btn-accent--hover {
	@apply hover:text-gray-50
  hover:bg-primary-pure;
}

// materil style
span.mat-button-wrapper {
	position: relative;
	top: 2px;
}
