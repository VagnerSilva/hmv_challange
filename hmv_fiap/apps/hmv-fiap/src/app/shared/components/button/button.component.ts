import {
	AfterViewInit,
	Attribute,
	ChangeDetectionStrategy,
	Component,
	ElementRef,
	Input,
	ViewEncapsulation,
} from "@angular/core"
import { ButtonType, ColorType } from "../types"

type className = string
type ColorsType = {
	[key: string]: className[]
}

export const BUTTON_COLORS: ColorsType = {
	primary: ["hmv-btn-primary", "hmv-btn-primary--hover"],
	accent: ["hmv-btn-accent", "hmv-btn-accent--hover"],
}

@Component({
	selector: "hmv-button",
	templateUrl: "./button.component.html",
	styleUrls: ["./button.component.scss"],
	encapsulation: ViewEncapsulation.None,
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ButtonComponent implements AfterViewInit {
	@Input() color: ColorType = "primary"
	@Input() disabled!: boolean
	@Input() ariaLabel!: string

	constructor(@Attribute("type") public type: ButtonType, private el: ElementRef<HTMLElement>) {
		if (this.type === null) {
			this.type = `submit`
		}
	}

	ngAfterViewInit(): void {
		this._changeColorButton()
	}

	private _changeColorButton() {
		const button: HTMLElement = this.el.nativeElement.children.item(0) as HTMLElement
		if (button) {
			button.classList.add(...BUTTON_COLORS[this.color])
		}
	}
}
