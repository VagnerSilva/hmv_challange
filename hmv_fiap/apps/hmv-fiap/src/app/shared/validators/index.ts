export class Validator {
	static msg = new Validator()

	REQUIRED = 'Campo obrigatório'

	MIN(value: number): string {
		return `Campo deve ter no mínimo ${value} caracteres`
	}
}
