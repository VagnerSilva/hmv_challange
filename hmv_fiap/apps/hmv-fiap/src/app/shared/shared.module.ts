import { CommonModule } from '@angular/common'
import { NgModule } from '@angular/core'
import { ReactiveFormsModule } from '@angular/forms'
import { RxReactiveFormsModule } from '@rxweb/reactive-form-validators'
import { ButtonRadioComponent } from './components/button-radio/button-radio.component'
import { ButtonComponent } from './components/button/button.component'
import { InputComponent } from './components/input/input.component'
import { SelectComponent } from './components/select/select.component'
import { TextareaComponent } from './components/textarea/textarea.component'
import { MaterialModule } from './material/material.module'

const components = [
	ButtonComponent,
	InputComponent,
	SelectComponent,
	TextareaComponent,
	ButtonRadioComponent,
]

const modules = [MaterialModule, ReactiveFormsModule]

@NgModule({
	declarations: [...components],
	imports: [CommonModule, ...modules],
	exports: [...components, ...modules, RxReactiveFormsModule],
})
export class SharedModule {}
