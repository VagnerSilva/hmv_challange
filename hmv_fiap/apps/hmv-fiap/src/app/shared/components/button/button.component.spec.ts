/* eslint-disable @nrwl/nx/enforce-module-boundaries */
import { SpectatorHost } from "@ngneat/spectator"
import { createHostFactory } from "@ngneat/spectator/jest"
import { SharedModule } from "@shared/shared.module"
import { axe } from "jest-axe"
import { ButtonComponent, BUTTON_COLORS } from "./button.component"

const BUTTON_NAME = `TESTE`
const cases = {
	simple: `<hmv-button>${BUTTON_NAME}</hmv-button>`,
	accent: `<hmv-button color="accent">${BUTTON_NAME}</hmv-button>`,
}

describe("ButtonComponent", () => {
	const createHost = createHostFactory({
		component: ButtonComponent,
		declareComponent: false,
		imports: [SharedModule],
	})

	function suw(host: SpectatorHost<ButtonComponent, unknown>) {
		return {
			button: host.query(`button`) as HTMLButtonElement,
		}
	}

	it("Should create component", () => {
		const host = createHost(cases.simple)
		const { button } = suw(host)

		expect(host.component.color).toBe(`primary`)
		expect(host.component.type).toBe(`submit`)

		expect(button.className).toContain(`primary`)
		expect(host.component).toBeTruthy()
		expect(button).toHaveText(BUTTON_NAME)
		expect(host.hostFixture).toMatchSnapshot()
	})

	it("Ensure that changes color when changing color property", () => {
		const host = createHost(cases.accent)
		const { button } = suw(host)

		host.component.ngAfterViewInit = jest.fn()
		host.component.ngAfterViewInit()

		expect(host.component.color).toBe("accent")
		expect(host.component.ngAfterViewInit).toHaveBeenCalledTimes(1)
		expect(button).toHaveText(BUTTON_NAME)
		expect(button).toHaveClass(BUTTON_COLORS["accent"])
		expect(host.hostFixture).toMatchSnapshot()
	})

	it("should verify component accessibility", async () => {
		const host = createHost(cases.simple)
		expect(await axe(host.hostFixture.nativeElement)).toHaveNoViolations()
	})
})
