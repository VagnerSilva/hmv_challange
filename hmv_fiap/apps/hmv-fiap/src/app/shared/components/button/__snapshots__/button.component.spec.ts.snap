// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent Ensure that changes color when changing color property 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <hmv-button
    color="accent"
  >
    <button
      class="mat-focus-indicator mat-raised-button mat-button-base _mat-animation-noopable hmv-btn-accent hmv-btn-accent--hover"
      mat-raised-button=""
      type="submit"
    >
      <span
        class="mat-button-wrapper"
      >
        TESTE
      </span>
      <span
        class="mat-ripple mat-button-ripple"
        matripple=""
      />
      <span
        class="mat-button-focus-overlay"
      />
    </button>
  </hmv-button>
</ng-component>
`;

exports[`ButtonComponent Should create component 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <hmv-button>
    <button
      class="mat-focus-indicator mat-raised-button mat-button-base _mat-animation-noopable hmv-btn-primary hmv-btn-primary--hover"
      mat-raised-button=""
      type="submit"
    >
      <span
        class="mat-button-wrapper"
      >
        TESTE
      </span>
      <span
        class="mat-ripple mat-button-ripple"
        matripple=""
      />
      <span
        class="mat-button-focus-overlay"
      />
    </button>
  </hmv-button>
</ng-component>
`;
